Cython/Utility/CythonFunction.c

```diff
@@ -535,25 +530,38 @@ __Pyx_CyFunction_get_is_coroutine_locked(__pyx_CyFunctionObject *op) {
         module = PyImport_ImportModuleLevelObject(PYIDENT("asyncio.coroutines"), NULL, NULL, fromlist, 0);
         Py_DECREF(fromlist);
         if (unlikely(!module)) goto ignore;
-        op->func_is_coroutine = __Pyx_PyObject_GetAttrStr(module, marker);
+        is_coroutine_value = __Pyx_PyObject_GetAttrStr(module, marker);
         Py_DECREF(module);
-        if (likely(op->func_is_coroutine)) {
-            return __Pyx_NewRef(op->func_is_coroutine);
+        if (likely(is_coroutine_value)) {
+            return is_coroutine_value;
         }
 ignore:
         PyErr_Clear();
     }
 
-    op->func_is_coroutine = __Pyx_PyBool_FromLong(is_coroutine);
-    return __Pyx_NewRef(op->func_is_coroutine);
+    return __Pyx_PyBool_FromLong(is_coroutine);
 }
 
 static PyObject *
 __Pyx_CyFunction_get_is_coroutine(__pyx_CyFunctionObject *op, void *context) {
     PyObject *result;
     CYTHON_UNUSED_VAR(context);
+    if (op->func_is_coroutine) {
+        return __Pyx_NewRef(op->func_is_coroutine);
+    }
```

Comment:
```suggestion
    __Pyx_BEGIN_CRITICAL_SECTION(op);
    if (op->func_is_coroutine) {
        result = __Pyx_NewRef(op->func_is_coroutine);
    }
    __Pyx_END_CRITICAL_SECTION();
    if (result) return result;
```

This is my proposed change. I'd rather avoid (technical) undefined behaviour so things are testable with thread sanitizer if we want to. That'd require atomics here. I don't think this is performance-critical enough to justify the extra complexity so a critical section seems the easiest approach.

Reply:
Isn't there a way to declare an "initialised once" value to thread-sanitizer? Once the if-test succeeds, the value is set and safe to be read by threads.

Reply:
I don't think there's a special way to declare it to thread sanitizer that concept to thread sanitizer. It normally just expects you to use atomics for the reads and write (which is fine, but the macros are a bit awkward).

Reply:
As a hackish form of a hack, we could inject the critical section only if TSan is used. :o]

https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual#suppressing-reports
"""
Exclude problematic code/test under TSan with `#if defined(__has_feature) && __has_feature(thread_sanitizer)`.
"""